korman/exporter/etlight.py

```diff
@@ -179,13 +181,75 @@ def _bake_static_lighting(self, bake, toggle):
                 elif key[0] == "vcol":
                     self._report.msg("{} Vertex Color(s) [H:{:X}]", len(value), hash(key[1:]), indent=1)
                     self._bake_vcols(value, key[1:])
+                    self._fix_vertex_colors(value)
                 else:
                     raise RuntimeError(key[0])
             inc_progress()
 
         # Return how many thingos we baked
         return sum(map(len, bake.values()))
 
+    def _fix_vertex_colors(self, blender_objects):
+        # Blender's lightmapper has a bug which allows vertices to "self-occlude" when shared between
+        # two faces. See here https://forum.guildofwriters.org/viewtopic.php?f=9&t=6576&p=68939
+        # What we're doing here is an improved version of the algorithm in the previous link.
+        # For each loop, we find all other loops in the mesh sharing the same vertex, which aren't
+        # separated by a sharp edge. We then take the brightest color out of all those loops, and
+        # assign it back to the base loop.
+        # "Sharp edges" include edges manually tagged as sharp by the user, or part of a non-smooth
+        # face, or edges for which the face angle is superior to the mesh's auto-smooth threshold.
+        # (If the object has an edge split modifier, well, screw you!)
+        for bo in blender_objects:
+            mesh = bo.data
+            bm = bmesh.new()
+            bm.from_mesh(mesh)
```

Comment:
I think I would feel better if this were a context manager to avoid the tedious and un-Pythonic manual resource management. So, you'd want something like a helper function:

```python
@contextmanager
def bmesh_edit_mesh(bl):
    bm = bmesh.new()
    bm.from_mesh(bl.data)
    try:
        yield bm
    except:
         pass
    else:
        bm.to_mesh(bl.data)
    finally:
        bm.free()
```

I'd probably use an exception to bail out of this context manager to avoid triggering the `bm.to_mesh` call, if absolutely needed - eg for line 211 `if light_vcol is None`.

Reply:
Right. Even better, with some tweaks I can simply reuse the existing `bmesh_from_object` from `helpers.py`, and avoid adding a new context manager.
Unfortunately, this also causes Blender to crash on `mesh.from_object(bl, bpy.context.scene)` the second time `bmesh_from_object` is called - for instance, when exporting a collision modifier. Seems specific to `from_object`, because `from_mesh` worked correctly earlier.
I don't have the means to debug the internals of Blender right now. If you have the time, [here is an example file](https://www.dropbox.com/s/jpev9ms2em8gvjq/TsoidahlPrad_279_kdebug_verylight.blend?dl=1) that causes the issue. Otherwise I'll see when I can setup a build environment to debug this.

Reply:
It wouldn't surprise me if some of the shenanigans that we're doing with collapsing the modifiers and using a temporary mesh are causing problems with `from_object`. I think it might be best to just go ahead and add a new helper with `from_mesh()`. It's painfully easy to crash Blender from Python, so I I'd prefer to stick with what we know works 😄

Reply:
Understood. Debugging crashes on the Python side already takes a lot of time, so to be honest I really wasn't looking forward to digging into Blender itself just for the sake of it :laughing:

Reply:
FWIW, I do have many working from source Blender builds for debugging, I just prefer to avoid actually doing that work. Especially when we have something we know works 😜